File: prism-kotlin-sdk/extras/src/commonMain/kotlin/io/iohk/atala/prism/kotlin/extras/AtalaOperationId.kt
package io.iohk.atala.prism.kotlin.extras

import com.benasher44.uuid.uuid4
import io.iohk.atala.prism.kotlin.crypto.SHA256Digest
import io.iohk.atala.prism.kotlin.crypto.util.BytesOps
import io.iohk.atala.prism.kotlin.protos.SignedAtalaOperation
import pbandk.encodeToByteArray
import kotlin.js.JsExport

@JsExport
data class AtalaOperationId(val digest: SHA256Digest) {
    fun value(): ByteArray = digest.value

    fun hexValue(): String = BytesOps.bytesToHex(value())

    companion object {
        fun of(atalaOperation: SignedAtalaOperation): AtalaOperationId {
            val hash = SHA256Digest.compute(atalaOperation.encodeToByteArray())
            return AtalaOperationId(hash)
        }

        fun random(): AtalaOperationId {
            val hash = SHA256Digest.compute(uuid4().toString().encodeToByteArray())
            return AtalaOperationId(hash)
        }

        fun fromHex(hex: String): AtalaOperationId {
            val hash = SHA256Digest.fromHex(hex)
            return AtalaOperationId(hash)
        }
    }
}

File: prism-backend/common/src/main/scala/io/iohk/atala/prism/connector/AtalaOperationId.scala
package io.iohk.atala.prism.connector

import com.google.protobuf.ByteString
import io.iohk.atala.prism.kotlin.crypto.Sha256Digest
import io.iohk.atala.prism.kotlin.crypto.Sha256
import io.iohk.atala.prism.protos.node_models
import io.iohk.atala.prism.utils.BytesOps

import java.util.UUID

class AtalaOperationId private (val digest: Sha256Digest) {
  def value: Vector[Byte] = digest.getValue.toVector

  def hexValue: String = BytesOps.bytesToHex(value)

  def toProtoByteString: ByteString = ByteString.copyFrom(value.toArray)

  override def equals(obj: Any): Boolean =
    obj match {
      case that: AtalaOperationId => this.digest.equals(that.digest)
      case _ => false
    }

  override def hashCode(): Int = digest.hashCode()

  override def toString: String = s"OperationId($hexValue)"
}

object AtalaOperationId {
  def of(atalaOperation: node_models.SignedAtalaOperation): AtalaOperationId = {
    val hash = Sha256.compute(atalaOperation.toByteArray)
    new AtalaOperationId(hash)
  }

  def random(): AtalaOperationId = {
    val hash = Sha256.compute(UUID.randomUUID().toString.getBytes())
    new AtalaOperationId(hash)
  }

  def fromVectorUnsafe(bytes: Vector[Byte]): AtalaOperationId = {
    val hash = Sha256Digest.fromBytes(bytes.toArray)
    new AtalaOperationId(hash)
  }

  def fromHexUnsafe(hex: String): AtalaOperationId = {
    val hash = Sha256Digest.fromHex(hex)
    new AtalaOperationId(hash)
  }
}
